validator/testcases/frameworks.py

```diff
@@ -0,0 +1,546 @@
+import json
+import re
+
+from validator.constants import LIBRARY_METADATA
+from validator.decorator import cached_property
+from validator.testcases import libraries
+from validator.version import Version
+
+
+def fake_version(files):
+    """
+    Returns a hash to act in place of a version number, given the contents
+    of a set of known files.
+    """
+    import hashlib
+    return hashlib.sha256("".join(files)).hexdigest()[:16]
+
+
+class MetaFramework(type):
+    def __new__(mcls, name, bases, dict_):
+        cls = super(MetaFramework, mcls).__new__(mcls, name, bases, dict_)
+        if cls.NAME:
+            cls.FRAMEWORKS.append(cls)
+        return cls
+
+
+class Framework(object):
+    """
+    Base class for framework definitions.
+    """
+
+    FRAMEWORKS = []
+
+    SAFE_FILES = ".jpg", ".ico", ".png", ".gif", ".txt"
+
+    CHECKED_FILES = re.compile(r"\.(jsm?|xul|xbl|x?html?)$", re.I)
+
+    __metaclass__ = MetaFramework
+
+    def __init__(self, err, xpi):
+        self.err = err
+        self.xpi = xpi
+
+        # Update certain metadata from the metadata JSON.
+        # Ideally this would also update based on data for the
+        # specific version of the framework, but there are
+        # bootstrapping issues.
+        for prop in "framework_files", "framework_prefixes":
+            if prop in self.framework_metadata:
+                setattr(self, prop.upper(),
+                        tuple(self.framework_metadata[prop]))
+
+    NAME = None
+    ACCEPTED = True
+
+    # Path of the JSON file within the XPI containing framework
+    # metadata. Not applicable to all frameworks.
+    METADATA_FILE = None
+
+    # Tuple of file paths which, if they exist, belong to the framework.
+    FRAMEWORK_FILES = ()
+
+    # Tuple of file path prefixes for which matching files, if they
+    # exist, belong to the framework.
+    FRAMEWORK_PREFIXES = ()
+
+    @cached_property
+    def ID(self):
+        """Internal ID of the framework."""
+        return self.NAME and self.NAME.lower()
+
+    def detect(self):
+        """Returns true if the add-on matches the given framework."""
+        return False
+
+    @property
+    def framework_metadata(self):
+        """
+        Returns the metadata for this framework, as defined in the
+        libraries metadata file if it exists, or a resonable
+        skeleton definition if it does not.
+        """
+        DEFAULT = {"versions": {}}
+
+        return LIBRARY_METADATA["frameworks"].get(self.ID, DEFAULT)
+
+    @cached_property
+    def framework_files(self):
+        """
+        Returns a list of files in the package which belong to the framework
+        in question. Files may or may not match known versions of the file.
+        """
+        return (filter(self.xpi.__contains__, list(self.FRAMEWORK_FILES)) +
```

Comment:
why the call to `list`, it says from the comment https://github.com/mozilla/amo-validator/pull/271/files#diff-247aeb318317091e4843b00ac0f88b12R60 that it's a tuple?